exonum/src/blockchain/service/srv.rs

```diff
@@ -0,0 +1,247 @@
+extern crate core;
+
+use serde_json::Value;
+
+use iron::prelude::*;
+use iron::Handler;
+use router::Router;
+
+use serde_json;
+
+use storage::{Fork, Snapshot};
+use messages::RawTransaction;
+use encoding::Error as MessageError;
+use crypto::Hash;
+use api::{ApiError, Api};
+use node::{TransactionSend, ApiSender};
+
+
+use super::{Transaction, ServiceContext, ApiContext, Service};
+use std::marker::PhantomData;
+
+/// TODO
+pub trait TransactionSet {
+    /// TODO
+    fn tx_from_raw(raw: RawTransaction) -> Result<Box<Transaction>, MessageError>;
+    /// TODO
+    fn tx_from_request(request: &mut Request) -> Result<Box<Transaction>, ApiError>;
+}
+
+#[macro_export]
+macro_rules! transaction_set {
+    ( $name:ident { $($tx:ident),* } ) => {
+        pub enum $name {
+        }
+
+        impl $crate::blockchain::TransactionSet for $name {
+            fn tx_from_raw(raw: RawTransaction) -> Result<Box<$crate::blockchain::Transaction>, $crate::encoding::Error> {
+                let message_type = raw.message_type();
+                $(
+                    if $tx::message_id() == message_type {
+                        return Ok(Box::new($tx::from_raw(raw)?));
+                    }
+                )*
+
+                return Err($crate::encoding::Error::IncorrectMessageType { message_type })
+            }
+
+            //TODO: reexport & $crate::
+            fn tx_from_request(request: &mut $crate::iron::Request) -> Result<Box<$crate::blockchain::Transaction>, $crate::api::ApiError> {
+                use $crate::iron::prelude::*;
+
+                #[derive(Deserialize, Clone)]
+                #[serde(untagged)] // :(
+                enum Any {
```

Comment:
I wonder whether it would make sense to implement this method akin to `tx_from_raw`, i.e., by initially parsing a `serde_json::Value` object and matching its `message_id` field. This would allow to avoid linear dependency of the parsing time on the number of transaction types. Would it be too expensive?

Reply:
Yeah, I plan to look into serialization closely, the `:(` here is for a reason :) We really should teach serde to use the value of message_id for choosing the variant. Now I believe this is even **incorrect** if two different transactions share the same set of fields. 

I don't think that literally producing a `json::Value` is the right approach (json object is much heavier than the corresponding message), but I am almost sure that there is *some* way to convince `serde` to do what we want it to do.

Reply:
Maybe extracting *only* the `message_id` field value and skipping all other data a-la [this serde example](https://serde.rs/stream-array.html) could work? 🤔

Reply:
Re: transactions with the same fields. We've discussed this before, and it seems that it would work correctly because a transaction deserialization logic expects a particular `message_id` value and fails otherwise. Still seems inefficient.